Importance sampling of OSL environment EDFs

lorentzo opened the pull request:
## Contribution of this PR

### Issue:
This is PR for issue #701 (Importance sampling of OSL environment EDFs).

### Summary:
* Implementation of `OSLEnvironmentSampler` which is using `OSLShadingSystem` and `ShaderGroup` for evaluating background shader. (**NEW:** instead of directly using `OSLShadingSystem` I am using `OSLShaderGroupExec`)
* Building of importance map inside of `importance_sampler` (instance of [ImageImportanceSampler](https://github.com/appleseedhq/appleseed/blob/3ac04302ca461107e5eacc70afa9a1a300645628/src/appleseed/foundation/math/sampling/imageimportancesampler.h) which is using `OSLEnvironmentSampler` as `sampler` in `ImageImportanceSampler::rebuild()`). This is done by modifying `build_importance_map()`.
* Modification of `OSLEnvironmentSampler::sample()`, `OSLEnvironmentSampler::evaluate()` and `OSLEnvironmentSampler::evaluate_pdf()` so it is using the importance_sampler.
* Some of my additional comments during solving this issue are available here: https://github.com/appleseedhq/appleseed/issues/701
* My work was inspired by:
  * [LatLongMapEnvironmentEDF](https://github.com/appleseedhq/appleseed/blob/027ef181f5c7bd7be88d79ecd69be10a35698b41/src/appleseed/renderer/modeling/environmentedf/latlongmapenvironmentedf.cpp)
  * [OSL's test render component: background](https://github.com/imageworks/OpenShadingLanguage/blob/master/src/testrender/background.h)

## Notes:
* I haven't used `OSLShaderGroupExec` class nor `ShadingContext` class for background shader evaluation since `OSLShaderGroupExec::execute_background()` is private. Instead I have written similar code for evaluation of background shader. (**NEW:** I am using `OSLShaderGroupExec` for evaluation of background shader and now `OSLShaderGroupExec::execute_background()` is made public)
* Size of the importance map is is hard-coded - but can be changed easily later (noted as TODO in code). Let me know how would you like it to be set. (**NEW:** I am using entity meta data for enabling user values and default values)
* Next to importance map size, there may be some additional optional input parameters (noted as TODO in code). 
* Next to those "TODO" code comments, there are some code comments that I find useful. Let me know if they should stay. 
* For the testing I have executed ["29 - osl environment reflection direct lighting.appleseed"](https://github.com/appleseedhq/appleseed/blob/582484d396ba4d7996582c2cb0f274606d848c70/sandbox/tests/test%20scenes/environment/29%20-%20osl%20environment%20reflection%20direct%20lighting.appleseed) and executed modification of ["17 - latlong environment map out-of-range luminance.appleseed"](https://github.com/appleseedhq/appleseed/blob/582484d396ba4d7996582c2cb0f274606d848c70/sandbox/tests/test%20scenes/environment/17%20-%20latlong%20environment%20map%20out-of-range%20luminance.appleseed) where OSL shaders are used. I can also add this test to PR if needed. 
  * I have some doubts about resulting test images. I see no difference in resulting images of my implementation of importance sampling of OSL environment EDFs and one that existed before. 
  * Also, test "17 - latlong environment map out-of-range luminance.appleseed" results in image where really bright pixels are visible - I don't see that in my modification of this test when using my implementation of iImportance sampling of OSL environment EDFs.

## Further work
* I am interested into testing this code. I will try to think of some goods tests. But let me know if you have some ideas so I can try them out.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +91,81 @@ namespace renderer
+          , m_height(height)
+          , m_rcp_width(1.0f / width)
+          , m_rcp_height(1.0f / height)
+          , m_osl_thread_info(osl_shading_system->create_thread_info())
```
I think it would be better to make OSLShaderGroupExec::execute_background public instead or repeating code.

lorentzo replied:
Making OSLShaderGroupExec::execute_background() public seems as a good choice for this problem. But all other OSLShaderGroupExec class methods are private so it seemed to me that I would "break" the consistency by making OSLShaderGroupExec::execute_background() public. 
So if you agree, I will make OSLShaderGroupExec::execute_background() public.

est77 replied:
Yes, make it public. It is better than reimplenting the code.

dictoon replied:
Do I understand correctly that this has yet to be done?

lorentzo replied:
Yes, this still has to be done. I am running some tests now it should be up today.

lorentzo replied:
Done. Thanks for the comment.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,7 +395,121 @@ namespace
+                    &renderer_services,
+                    texture_system.get()));
+
+            // Create shader compiler.
```
The shader should have already been compiled and optimized here. You don't need to do it again.

lorentzo replied:
OK, I will remove it then. 
For satisfying my curiosity: which kind of optimization is made to shaders?

est77 replied:
LLVM optimization passes and JIT compiling to x64 assembly.

lorentzo replied:
Done, thanks for the explanation.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -31,26 +31,52 @@
 #include <cassert>
 #include <cstring>
 #include <string>
+#include <memory>
```
Order alphabetically.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+    class OSLEnvironmentSampler
+    {
+      public:
+
```
Remove blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
 
 namespace
 {
+
```
Remove blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+      public:
+
+        OSLEnvironmentSampler(
+            OSLShadingSystem* osl_shading_system,
```
Align arguments vertically.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+
+            OSL::ShaderGlobals sg;
+
+            memset(&sg, 0, sizeof(OSL::ShaderGlobals));
```
`std::memset()`

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+
+            importance = luminance(payload);
+        }
+
```
Remove one blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+
+
+      private:
+
```
Remove blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+
+      private:
+
+        ShaderGroup* m_shader_group;
```
Align members vertically.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+          : m_osl_shading_system(osl_shading_system)
+          , m_shader_group(shader_group)
+          , m_arena(arena)
+          , m_width(width)
```
It looks like you don't need `m_width` or `m_height`, you can remove them.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +89,81 @@ namespace renderer
+
+        ShaderGroup* m_shader_group;
+        OSLShadingSystem* m_osl_shading_system;
+        Arena& m_arena;
```
Not needed? Only keep what is really needed.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -99,12 +204,23 @@ namespace
             OnFrameBeginRecorder&   recorder,
             IAbortSwitch*           abort_switch) override
         {
+
```
Remove redundant blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -99,12 +204,23 @@ namespace
             if (!EnvironmentEDF::on_frame_begin(project, parent, recorder, abort_switch))
                 return false;
-
+ 
```
Remove trailing space.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -99,12 +204,23 @@ namespace
             m_shader_group =
                 static_cast<ShaderGroup*>(m_inputs.get_entity("osl_background"));
 
+            if(m_importance_sampler.get() == nullptr || m_shader_group->get_version_id() != m_shader_group_version_id){
```
Replace `m_importance_sampler.get() == nullptr` by `!m_importance_sampler`.

Insert space after `if`, move opening brace to its own line.

You can find our coding guidelines on the wiki:
https://github.com/appleseedhq/appleseed/wiki/Coding-Philosophy-and-Guidelines

dictoon replied:
Everything in this method almost certainly needs to be moved to `on_render_begin()`.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -99,12 +204,23 @@ namespace
+                // Build importance map only if this environment EDF is the active one.
+                if (project.get_scene()->get_environment()->get_uncached_environment_edf() == this)
+                    build_importance_map(project, abort_switch);
+            
```
Remove redundant blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -124,30 +240,75 @@ namespace
+            shift_angles(theta, phi, m_theta_shift, m_phi_shift);
+
+            // Compute the local space emission direction.
+            const float cos_theta = cos(theta);
```
Add `std::` qualifiers to all math functions (would also appreciate if you could fix the code you copied from).

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -124,30 +240,75 @@ namespace
             evaluate_osl_background(shading_context, local_outgoing, value);
         }
 
+        // Calculates value and probability, therefore importance map is needed.
```
This comment would be best placed right before the line

    if (!m_importance_sampler)

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,7 +393,109 @@ namespace
+
+            // TODO:
+            // Find importance map width and height from m_inputs (see LatLongMapEnvironmentEDF).
+            m_importance_map_width = 1024;
```
You should use the minimum of the importance map's width and some user value with a reasonable default (like 1024).

You can retrieve the user value from the entity's parameter (not the inputs that are bound connections such as colors and textures). You can check how this is done in numerous other entities.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,7 +393,109 @@ namespace
+                new ImageImportanceSamplerType(
+                    m_importance_map_width,
+                    m_importance_map_height));
+
```
Remove redundant blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -235,4 +548,4 @@ auto_release_ptr<EnvironmentEDF> OSLEnvironmentEDFFactory::create(
 }
 
-}   // namespace renderer
+}   // namespace renderer
```
End file with newline character.

lorentzo commented:
@dictoon thank you for detailed code review. I have fixed everything you asked. I have two doubts for now:
* moving `build_importance_map()` from `on_frame_begin()` to `on_render_begin()` as you proposed in https://github.com/appleseedhq/appleseed/pull/2784#discussion_r386053142 results in segmentation error while calling `m_osl_shadergroup_exec.execute_background()` (this function is needed for building the importance map). After some experiments I think I have figured out what might be the problem. I think that when `on_render_begin()` is called, OSL shaders haven't been configured yet (path, headers etc.). While when `build_importance_map()` is in `on_frame_begin()`, no problem occurs. That is because `on_frame_begin()` is called after the OSL shaders have been configured. I will discuss this further on discord.
* Do you require some additional test scenes? For example I have modified some test scenes for OSL environment where metadata is used.

lorentzo commented:
OK, now all commits have been squashed. No conflicts. Buildable. 


Only this is left to consider: https://github.com/appleseedhq/appleseed/pull/2784#issuecomment-593100136. 

I have discussed this problem in Discord #dev, here: https://discordapp.com/channels/430063582777049089/430064067873472526/683677424088514571.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +86,53 @@ namespace renderer
+    {
+      public:
+        OSLEnvironmentSampler(
+            ShaderGroup*          shader_group,
```
make const

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +86,53 @@ namespace renderer
+      public:
+        OSLEnvironmentSampler(
+            ShaderGroup*          shader_group,
+            OSLShaderGroupExec&   osl_shadergroup_exec,
```
make this one const too

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -76,11 +142,21 @@ namespace
+
+            // TODO:
+            // Add optional values from m_inputs (see LatLongMapEnvironmentEDF).
+            m_phi_shift = deg_to_rad(m_params.get_optional<float>("horizontal_shift", 0.0f));
```
Rotating the environment is something that can be easily done in an OSL shader if needed and that is the place where it should go. I prefer that we don't handle it here.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -76,11 +142,21 @@ namespace
+            m_phi_shift = deg_to_rad(m_params.get_optional<float>("horizontal_shift", 0.0f));
+            m_theta_shift = deg_to_rad(m_params.get_optional<float>("vertical_shift", 0.0f));
+            m_importance_map_width = m_params.get_optional<size_t>("importance_map_width", 1024);
+            m_importance_map_height = m_params.get_optional<size_t>("importance_map_height", 512);
```
We can always fix height to width / 2. It does not need to be a param.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -124,30 +224,75 @@ namespace
+            assert(probability >= 0.0f);            
         }
 
+        // Calculates only value, therefore importance map is not needed.
```
This comment is not needed.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -223,6 +513,76 @@ DictionaryArray OSLEnvironmentEDFFactory::get_input_metadata() const
                     .insert("shader_group", "Shader Groups"))
             .insert("use", "optional"));
 
+    metadata.push_back(
```
Remove shift params. We should do this inside the OSL shader if needed.

est77 commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -223,6 +513,76 @@ DictionaryArray OSLEnvironmentEDFFactory::get_input_metadata() const
+
+    metadata.push_back(
+        Dictionary()
+            .insert("name", "importance_map_width")
```
Maybe important_map_size?. We can then set width = size and height = size / 2.

lorentzo replied:
Agree. Done it like that.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +379,107 @@ namespace
+            assert(m_importance_sampler);
+
+            // Compute the probability density of this sample in the importance map.
+            size_t m_importance_map_width = m_importance_map_size;
```
Use `std::size_t` in new code.

lorentzo replied:
Done to all.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -223,6 +516,25 @@ DictionaryArray OSLEnvironmentEDFFactory::get_input_metadata() const
+                    .insert("type", "soft"))
+            .insert("default", "1024")
+            .insert("use", "optional")
+            .insert("help", "Importance map size. width = size, height = size / 2"));
```
Nice! I suggest tweaking it a bit (replaced period by colon, change spacing):

    "Importance map size: width = size, height = size/2"

lorentzo replied:
Agree, done.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +379,107 @@ namespace
+
+            // Create OSL shadergroup exec.
+            Arena arena;
+
```
Remove blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +379,107 @@ namespace
+            // Create OSL shading system.
+            auto_release_ptr<OIIOTextureSystem> texture_system(
+                OIIOTextureSystemFactory().create());
+
```
Remove blank line.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +379,107 @@ namespace
+                OIIOTextureSystemFactory().create());
+
+            RendererServices renderer_services(project, texture_system.ref());
+
```
Remove blank line.

The idea is that the comment "Create OSL shading system" really applies to the whole block so it shouldn't be interrupted by blank lines.

lorentzo replied:
I understand. Done.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -76,11 +151,17 @@ namespace
+
+            // TODO:
+            // Add optional values from m_inputs (see LatLongMapEnvironmentEDF).
+            m_importance_map_size = m_params.get_optional<size_t>("importance_map_size", 1024);
```
I think this would be best done in `on_render_begin()`.

lorentzo replied:
Agree. Done.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -76,11 +151,17 @@ namespace
         {
             m_inputs.declare("osl_background", InputFormatEntity, "");
+
+            // TODO:
```
I'm not sure to understand this TODO.

lorentzo replied:
Just a hint to add any additional parameters if needed in the future. But I have removed it because it sounds confusing. It is actually clear without a comment.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -63,10 +86,62 @@ namespace renderer
+        const ShaderGroup*          m_shader_group;
+        const OSLShaderGroupExec&   m_osl_shadergroup_exec;
+
+        const float           m_rcp_width;
```
Align members.

lorentzo replied:
Done in all.

dictoon commented:
To answer your previous points:

> Moving build_importance_map() from on_frame_begin() to on_render_begin() as you proposed in #2784 (comment) results in segmentation error while calling m_osl_shadergroup_exec.execute_background() (this function is needed for building the importance map). After some experiments I think I have figured out what might be the problem. I think that when on_render_begin() is called, OSL shaders haven't been configured yet (path, headers etc.). While when build_importance_map() is in on_frame_begin(), no problem occurs. That is because on_frame_begin() is called after the OSL shaders have been configured. I will discuss this further on discord.

That's interesting. I haven't checked exactly what's going on but I trust your analysis. What we could do to prevent this PR from stalling is keep this in `on_frame_begin()` for now, with a comment explaining the problem, then investigate how to fix the problem in a second time.

The impact of rebuilding the importance map in `on_frame_begin()` is low since you check if the shader group version has changed.

> Do you require some additional test scenes for this issue? For example I have modified some test scenes for OSL environment where metadata is used.

It would be nice to have at least one scene in the test suite that checks that this code works correctly.

lorentzo commented:
> To answer your previous points:
> 
> > Moving build_importance_map() from on_frame_begin() to on_render_begin() as you proposed in #2784 (comment) results in segmentation error while calling m_osl_shadergroup_exec.execute_background() (this function is needed for building the importance map). After some experiments I think I have figured out what might be the problem. I think that when on_render_begin() is called, OSL shaders haven't been configured yet (path, headers etc.). While when build_importance_map() is in on_frame_begin(), no problem occurs. That is because on_frame_begin() is called after the OSL shaders have been configured. I will discuss this further on discord.
> 
> That's interesting. I haven't checked exactly what's going on but I trust your analysis. What we could do to prevent this PR from stalling is keep this in `on_frame_begin()` for now, with a comment explaining the problem, then investigate how to fix the problem in a second time.
> 
> The impact of rebuilding the importance map in `on_frame_begin()` is low since you check if the shader group version has changed.

I would suggest opening an issue: "Investigate a way to add build importance map inside of  OSLEnvironmentEDF::on_render_begin() instead of OSLEnvironmentEDF::on_frame_begin()". I would reference this PR and provide analysis that I have done. Do you agree?

> > Do you require some additional test scenes for this issue? For example I have modified some test scenes for OSL environment where metadata is used.
> 
> It would be nice to have at least one scene in the test suite that checks that this code works correctly.

The test suite already contains a test scene that uses/test the code I have created, here: https://github.com/appleseedhq/appleseed/blob/master/sandbox/tests/test%20scenes/environment/29%20-%20osl%20environment%20reflection%20direct%20lighting.appleseed.
I will just add one more with the explicit parameter for the importance map size.

lorentzo commented:
OK, now all the required changes are done. PR is squashed and rebased on the master. It is ready for merging. Summary:
* Added importance sampling of OSL environment EDF.
* Added test scene for importance sampling of OSL environment EDF, new HDR image for the test and resulting image of the test.
* Refactoring of `latlongmapenvironmentedf.cpp`  and `oslshadergroupexec.h`

Further work. Opening issue: "Investigate a way to add build importance map inside of OSLEnvironmentEDF::on_render_begin() instead of OSLEnvironmentEDF::on_frame_begin()"

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +377,104 @@ namespace
+            assert(m_importance_sampler);
+
+            // Compute the probability density of this sample in the importance map.
+            std::size_t m_importance_map_width = m_importance_map_size;
```
Don't use `m_` prefix for local variables.

Make const.

lorentzo replied:
Done for all.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -156,19 +304,66 @@ namespace
 
         float evaluate_pdf(
-            const Vector3f&         outgoing) const override
+            const Vector3f&   outgoing) const override
```
Fits on one line.

lorentzo replied:
Done.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -141,13 +287,15 @@ namespace
             evaluate_osl_background(shading_context, local_outgoing, value);
         }
 
+        
```
Revert.

lorentzo replied:
Done.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +377,104 @@ namespace
+            Stopwatch<DefaultWallclockTimer> stopwatch;
+            stopwatch.start();
+
+            std::size_t m_importance_map_width = m_importance_map_size;
```
Same here, remove `m_` and make const.

lorentzo replied:
Done for all.

dictoon commented on src/appleseed/renderer/modeling/environmentedf/oslenvironmentedf.cpp:
```diff
@@ -182,11 +377,104 @@ namespace
+            }
+        }
+
+        ShaderGroup*                                  m_shader_group;
```
Move data members before methods.

lorentzo replied:
Done.

lorentzo commented:
HDR map `colorful_studio_2k.hdr` is taken from https://hdrihaven.com/ which states:

> All HDRIs are licenced as CC0 and can be downloaded instantly, giving you complete freedom.

dictoon commented:
Thanks for your contribution!
